### ISSUE TITLE
refactor Addressbook proposal

### ISSUE THREAD

**Issue by NodeGuy** [cosmos/voyager#1195]
The main process [is horribly complex](https://github.com/cosmos/voyager/issues/856).

Addressbook is a class in the main process that discovers and keeps a stateful collection of node addresses.  A node is chosen randomly from the pool when starting Gaia Lite.

The usage of the address book is essentially as follows:

1.  Load seed nodes from the network's `config.toml` file.
2.  Load more nodes that were saved to disk during the previous session.
3.  Filter out nodes that don't respond to a simple query.
4.  Discover new nodes by asking a node for its peers.
5.  Save nodes to disk for the next session.

I propose that we replace this complex, stateful class with a single stateless function.  The function will:

* take a set of seed nodes, discover new ones, and return a combined set of nodes known to respond
* combine the aliveness query with peer discovery to reduce the amount of network traffic
* make the network queries in parallel for increased performance
* not talk to the disk to eliminate the need to mock file system calls in tests

**Comment by jbibla** [comment]
this is excellent. really appreciate the clarity and the walkthrough. i didn't know anything about the address book - and now i do!

👍 👍

**Comment by NodeGuy** [comment]
Closing in favor of https://github.com/cosmos/voyager/issues/1291.

**Comment by faboweb** [comment]
@NodeGuy is the refactor not useful for #1291?

**Comment by NodeGuy** [comment]
> skip peer detection
> always connect to one of the seed nodes (full node)

According to my read of this, not only is the refactor not useful but neither is Addressbook any more.  Why keep an address book of peer nodes if we're connecting to only one node?  Am I understanding this correctly?

**Comment by faboweb** [comment]
Right. So should we archive the addressbook code then? In the future we probably want to have this again.

**Comment by NodeGuy** [comment]
I agree we'll probably want to have it again and I don't see any harm in leaving it as it is since it can handle the case of a single node already.
